dev/TeachingTip/TeachingTip.cpp

```diff
@@ -0,0 +1,1540 @@
+            {
+                return winrt::TeachingTipPlacementMode::TopEdgeAlignedLeft;
+            }
+            else if (bottomRightAvailable)
+            {
+                return winrt::TeachingTipPlacementMode::BottomEdgeAlignedRight;
+            }
+            else if (bottomLeftAvailable)
+            {
+                return winrt::TeachingTipPlacementMode::BottomEdgeAlignedLeft;
+            }
+            else if (rightTopAvailable)
+            {
+                return winrt::TeachingTipPlacementMode::RightEdgeAlignedTop;
+            }
+            else if (rightBottomAvailable)
+            {
+                return winrt::TeachingTipPlacementMode::RightEdgeAlignedBottom;
+            }
+            else if (leftTopAvailable)
+            {
+                return winrt::TeachingTipPlacementMode::LeftEdgeAlignedTop;
+            }
+            else if (leftBottomAvailable)
+            {
+                return winrt::TeachingTipPlacementMode::LeftEdgeAlignedBottom;
+            }
+        }
+    }
+    //the teaching tip wont fit anywhere, just return top.
+    return winrt::TeachingTipPlacementMode::Top;
+}
+
+
+void TeachingTip::EstablishShadows()
+{
+#ifdef USE_INSIDER_SDK
+    if (SharedHelpers::IsThemeShadowAvailable())
+    {
+#ifdef _DEBUG
+        auto beakShadow = winrt::Windows::UI::Xaml::Media::ThemeShadow{};
+        beakShadow.Receivers().Append(m_target.get());
+        m_beakPolygon.get().Shadow(beakShadow);
+        m_beakPolygon.get().Translation(winrt::float3{ m_beakPolygon.get().Translation().x, m_beakPolygon.get().Translation().y, m_beakElevation });
+#endif
+
+        auto contentShadow = winrt::Windows::UI::Xaml::Media::ThemeShadow{};
+        contentShadow.Receivers().Append(m_shadowTarget.get());
+        m_rootGrid.get().Shadow(contentShadow);
+        m_rootGrid.get().Translation(winrt::float3{ m_rootGrid.get().Translation().x, m_rootGrid.get().Translation().y, m_contentElevation });
+    }
+#endif
+}
+
+void TeachingTip::OnPropertyChanged(
+    const winrt::DependencyObject& sender,
+    const winrt::DependencyPropertyChangedEventArgs& args)
+{
+    if (args.Property() == s_AttachProperty)
+    {
+        OnAttachPropertyChanged(sender, args);
+    }
+    else
+    {
+        winrt::get_self<TeachingTip>(sender.as<winrt::TeachingTip>())->OnPropertyChanged(args);
+    }
+}
+
+void TeachingTip::OnAttachPropertyChanged(
+    const winrt::DependencyObject& sender,
+    const winrt::DependencyPropertyChangedEventArgs& args)
+{
+    auto oldTip = unbox_value<winrt::TeachingTip>(args.OldValue());
+    auto newTip = unbox_value<winrt::TeachingTip>(args.NewValue());
+
+    if (oldTip == newTip)
+    {
+        return;
+    }
+
+    winrt::TeachingTip::SetAttach(nullptr, oldTip);
+    winrt::TeachingTip::SetAttach(sender.try_as<winrt::UIElement>(), newTip);
+}
+
+void TeachingTip::SetAttach(const winrt::UIElement& element, const winrt::TeachingTip& teachingTip)
+{
+    MUX_ASSERT(teachingTip);
```

Comment:
Consider using gsl::not_null. We haven't started using it, but now is as good a time as any. I believe you'd use it in the parameter type like:

```c++
    void TeachingTip::SetAttach(const winrt::UIElement& element, const gsl::not_null<winrt::TeachingTip>& teachingTip)
```

Reply:
This *might* not work, there is [an open issue](https://github.com/Microsoft/GSL/issues/89) that hints it might not be fully supported. But worth a try.

Reply:
I'll wait until this is resolved.

---
In reply to: [239326611](https://github.com/Microsoft/microsoft-ui-xaml/pull/32#discussion_r239326611) [](ancestors = 239326611)